### PR TITLE
fix: bugs and improvements in link checker

### PR DESCRIPTION
**Pull request by nilubava**
Fixes:
- #4414
- #4084 
- #2512 
- #2500 
- #2650 
- #2734 
- #3330 
- #2723 
- #4628

### CODE REVIEW

**Comment by janpio** [comment]
Nice:

```
warning /300-guides/200-deployment/110-deployment-guides/600-deploying-migrations-from-a-local-environment contains links with trailing slashes: ./ - Please remove trailing slashes.
warning /200-concepts/100-components/01-prisma-schema/06-relations/420-relation-mode contains links with domain name prisma.io: https://www.prisma.io/docs/guides/database/using-prisma-with-planetscale#how-to-emulate-relations-in-prisma-client - Please use relative urls instead.
```
I think "use relative urls instead" is not correct though. We don't want relative links necessarily, but links without the domain that are still fully qualified in the path and not relative to the current page. Not sure how those are called actually.


Does this really cover https://github.com/prisma/docs/issues/2500 or https://github.com/prisma/docs/issues/2650 though? Would those be already "applied" when the linkchecker runs

**Comment by janpio** [comment]
A few things that might be false positives:
```
warning /200-concepts/100-components/01-prisma-schema/04-data-model contains links with trailing slashes: ./ - Please remove trailing slashes.
```
Does that actually cause a link with a trailing slash and not just the URL of the current page?

```
warning /200-concepts/100-components/02-prisma-client/058-transactions contains links with domain name prisma.io: https://www.prisma.io/dataguide/intro/database-glossary#isolation-levels - Please use relative urls instead.
```
Data guide should be ok - only docs internal links are problematic as it makes the previews not work.

Also:
```
Error: 0 broken (or redirected) internal links and 42 broken anchors found
```

Do these 2 new categories of broken links (trailing slash, and external URLs for prisma.io) count to "broken anchors"?

**Comment by nilubava** [comment]
@janpio 
> I think "use relative urls instead" is not correct though: 

I can change the wording to "Remove the domain name prisma.io from the link"?

> A few things that might be false positives

I can exclude such links from the warnings.

> Data guide should be ok 

I included dataguide and blog since they can also be written as relative links without domain names. 

> Do these 2 new categories of broken links (trailing slash, and external URLs for prisma.io) count to "broken anchors"?

Nope, they are caught in another plugin which transforms the urls (`gatsby-remark-to-absolueturl`). The broken anchors and links are caught in `gatsby-remark-check-links-numberless`. Hence the numbers are not for all the broken links

**Comment by janpio** [comment]
> I can change the wording to "Remove the domain name prisma.io from the link"?

👍 

> I included dataguide and blog since they can also be written as relative links without domain names.

I would only apply that to things that will work in the preview deployment.

> Nope, they are caught in another plugin which transforms the urls (gatsby-remark-to-absolueturl). The broken anchors and links are caught in gatsby-remark-check-links-numberless. Hence the numbers are not for all the broken links

Ahh, that makes sense - I did not actually look at the code. Would they also break the build? Maybe we should add a summary reporting that is output as well for them so it is easier to see what is going on? I had to look quite closely to actually notice these in the output.

**Comment by nilubava** [comment]
@janpio 
> Maybe we should add a summary reporting that is output as well

Yes, that makes sense. I will add this

File: content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/01-how-to-upgrade.mdx [hunk 1]
```diff
@@ -17,7 +17,7 @@ The upgrade documentation consists of several pages, here's an overview of how t
 
 In addition to these two pages, there are various _practical guides_ that walk you through an example scenario of the upgrade process:
 
-- [Upgrading the Prisma layer](/guides/upgrade-guides/upgrade-from-prisma-1/upgrading-the-prisma-layer-postgres): No matter what your Prisma 1 setup looks like, you should **always start your upgrade process by following this guide**.
+- [Upgrading the Prisma layer](/guides/upgrade-guides/upgrade-from-prisma-1/upgrading-the-prisma-layer-postgres-postgres-postgres-postgres-postgres): No matter what your Prisma 1 setup looks like, you should **always start your upgrade process by following this guide**.
```
Comment (janpio) [inline content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/01-how-to-upgrade.mdx]: This look broken 😆

Comment (nilubava) [reply content/300-guides/300-upgrade-guides/800-upgrade-from-prisma-1/01-how-to-upgrade.mdx]: Ha! that was a logic glitch 😆 I have fixed those now

**Comment by nilubava** [comment]
@janpio Looks like it can be merged now. I have fixed the redirects using the new script and it works as expected. I compared with Screaming Frog 308s and looks like it caught everything.

Should we make it a warning and not throw an error if links fail? This might help the doc writers to fix the links when they have time and not block further PRs (due to link errors). Once we have confidence that we have fixed all the links, we can think about making it an error again. WDYT?

**Pull request merged**